main/src/main/java/org/mobilitydata/gtfsvalidator/util/CalendarUtil.java

```diff
@@ -0,0 +1,159 @@
+        Set<LocalDate> removedDays = new HashSet<>();
+        for (GtfsCalendarDate calendarDate : calendarDates) {
+            (calendarDate.exceptionType() ==
+                    GtfsCalendarDateExceptionType.SERVICE_ADDED
+                    ? addedDays
+                    : removedDays)
+                    .add(calendarDate.date().getLocalDate());
+        }
+
+        return new ServicePeriod(serviceStart, serviceEnd, weeklyPattern,
+                addedDays, removedDays);
+    }
+
+    /**
+     * Builds a service id to {@code ServicePeriod} mapping using the given
+     * <i>calendar.txt</i> and <i>calendar_dates.txt</i> tables. <p> If either
+     * table is missing, a empty container should be passed in.
+     *
+     * @param calendarTable     the <i>calendar.txt</i> table container, may be
+     *                          empty
+     * @param calendarDateTable the <i>calendar_dates.txt</i> table container,
+     *                          may be empty
+     * @return mapping from service id to {@link ServicePeriod} object
+     */
+    public static Map<String, ServicePeriod> buildServicePeriodMap(
+            @Nonnull GtfsCalendarTableContainer calendarTable,
+            @Nonnull GtfsCalendarDateTableContainer calendarDateTable) {
+        Preconditions.checkNotNull(calendarTable);
+        Preconditions.checkNotNull(calendarDateTable);
+
+        Map<String, ServicePeriod> servicePeriods = new HashMap<>();
+        for (GtfsCalendar calendar : calendarTable.getEntities()) {
+            servicePeriods.put(
+                    calendar.serviceId(),
+                    createServicePeriod(calendar, calendarDateTable.byServiceId(
+                            calendar.serviceId())));
+        }
+        for (List<GtfsCalendarDate> calendarDates :
+                Multimaps.asMap(calendarDateTable.byServiceIdMap()).values()) {
+            if (!servicePeriods.containsKey(calendarDates.get(0).serviceId())) {
+                servicePeriods.put(calendarDates.get(0).serviceId(),
+                        createServicePeriod(null, calendarDates));
+            }
+        }
+        return servicePeriods;
+    }
+
+    /**
+     * Converts a map {service id -> ServicePeriod} to map {service id -> set of dates}.
+     *
+     * @param servicePeriods mapping from service id to {@link ServicePeriod}
+     * @return mapping from service id to a set of included days
+     */
+    public static Map<String, SortedSet<LocalDate>> servicePeriodToServiceDatesMap(
+            @Nonnull Map<String, ServicePeriod> servicePeriods) {
+        Map<String, SortedSet<LocalDate>> serviceDates = new HashMap<>();
+        for (Map.Entry<String, ServicePeriod> kv : servicePeriods.entrySet()) {
+            serviceDates.put(kv.getKey(), kv.getValue().toDates());
+        }
+        return serviceDates;
+    }
+
+    /**
+     * Finds the first intersecting date in the given sets, if any. Returns {@code Optional.empty()} if there is no
```

Comment:
Suggest adding big O notation of method runtime in Javadocs

Reply:
Done.